content/python/concepts/files/terms/context-managers/context-managers.md

```diff
@@ -0,0 +1,139 @@
+
+The below code exemplifies the class-based approach applied to a HR department sending automated yearly bonus notifications to employees. It's then quite easy to combine this context manager together with a [dictionary](https://www.codecademy.com/resources/docs/python/dictionaries) that contains all employees' records, loop through it, and have all the letters ready in no time.
+
+```py
+class bonus:
+  def __init__(self, employee_first,employee_id, salary, percentage):
+    self.first = employee_first
+    self.id = employee_id
+    self.salary = salary
+    self.pct = percentage
+    self.letter = open(str(self.id)+'_bonus_notice.txt','w')
+    self.thanks_message = open('thanks_message.txt', 'r')
+    self.conclusion = open('conclusion.txt','r')
+
+  def __enter__(self):
+    self.letter.write('Dear '+self.first +'!\n\n')
+    self.letter.write(self.thanks_message.read())
+    self.letter.write('\nYou will receive in addition to your December salary a yearly bonus that amounts to $'\
+    +str(round(self.salary*(self.pct/100)))+'.\n')
+
+  def __exit__(self,*exc):
+    self.letter.write(self.conclusion.read())
+    self.thanks_message.close()
+    self.conclusion.close()
+    self.letter.close()
+    return True
+
+with bonus('John',3789,100000,5) as john_smith_bonus:
+  print('Letter printed!')
+```
+
+Result:
+
+```md
+Dear John!
+
+We value all our employees and are committed to their growth and well-being. We would like to thank you for your amazing work throughout this whole year.
+You will receive in addition to your December salary a yearly bonus that amounts to $5000.
+Again, allow us to thank you for all your efforts and to wish you all the best, for you and your family.
+
+Sincerely yours,
+Your HR department
+```
+
+### Decorator-based approach
+
+In this approach, rather than a class is defined a [generator](https://www.codecademy.com/resources/docs/python/generators) function (i.e. a function performing a _yield_ and not a _return_). This function should be _decorated_ using `@contextmanager` from the built-in module `contextlib`:
+
+```py
+from contextlib import contextmanager
+```
+
+A decorator-based context manager is also using the user input in the form of parameters provided at the instantiation. It follows the syntax _try-except-finally_, _try_ being the part where the _yield_ statement is and the core operations are done, _except_ the exceptions handling part, and _finally_ the part where the resources are closed, essentially.
+
+The below code shows an implementation of a decorator-based context manager thanks to which interview proposals are sent to pre-selected candidates. It is also possible (but not mandatory) to introduce some personalized elements in the indented block. If doing so this element will be placed just before the timeslot proposal.
```

Comment:
````suggestion
### Decorator-Based Approach

In decorator-based approach, a [generator](https://www.codecademy.com/resources/docs/python/generators) function (i.e., a function performing a `_yield_ ` and not a `_return_`) is defined rather than a class. This function is _decorated_ using `@contextmanager` from the built-in module `contextlib`:

```pseudo
from contextlib import contextmanager
```

A decorator-based context manager also uses the user input in the form of parameters provided at the instantiation. It follows the syntax `_try-except-finally_`, where `_try_` being the part where the `_yield_` statement is and the core operations are done, `_except_` being the part where the exceptions are handled and `_finally_` being the part where the resources are closed.

The below code shows an implementation of a decorator-based context manager thanks to which interview proposals are sent to pre-selected candidates. It is also possible (but not mandatory) to introduce some personalized elements in the indented block. Doing so places this element just before the timeslot proposal:
````